### ISSUE TITLE
Packet/NBT Tag Size issue caused by Velocity

### ISSUE THREAD

**Issue by LandonDeam** [PaperMC/Velocity#1037]
**Steps to reproduce packet size issue:**
Run a 1.19.2 Fabric server with the [Dogcraft modpack](https://modrinth.com/modpack/dogcraft-official) under a Velocity server. Attempt to connect. Be met with the following (_server side_): https://mclo.gs/IHaC5YP

**Steps to reproduce NBT tag size issue:**
Run a 1.19.2 Fabric server with the mods Chipped, Consistency+, Continuity, Create, and Create Deco (all of which are in the aforementioned modpack) under a Velocity server. Attempt to connect. Be met with the following (_client-side_): https://mclo.gs/IUUz5pu

**My best guess as to the cause:** Velocity probably shares/uses the same built-in, hard-coded values for packet sizes and NBT tags as vanilla, and as soon as it sees something go over these limits, it throws up an error and disconnects the player, even if the player would have not been kicked by the Minecraft server itself. 
_Note: This is also pure speculation, so feel free to look for other explanations if this is not the case._

**Additional Info:** These errors do not occur when running the server bare, without Velocity. The root cause would seem to be Velocity somehow mishandling data that is too large, which tends to happen with the listed mods for some reason (which is strange, considering they basically all just add blocks). I also _would_ use the Connectivity mod and call it a day, but it seems to have some sort of conflict with at least one of the other mods in the above pack (the reason I am reporting this in the first place), so that solution is out the window for the time being.

**Comment by LandonDeam** [comment]
I'm currently suspecting it's something that occurs between lines 68 and 81 in the MinecraftDecoder class, as this would be the time for Velocity to so sanity checks on the buffer (and throw errors similar to the one described), and this function is only called by another function that, seemingly, would not have any effect on the packet/buffer. That is, of course, unless ByteBuf has a maximum buffer size set somewhere, and I'm just unfamiliar with the class.

**Comment by isitreallyalive** [comment]
I am also having issues with NBT tag sizes. I am running [Prominence II](https://www.curseforge.com/minecraft/modpacks/prominence-2-fabric). However, my issues are on the server side. This is what my error looks like: https://mclo.gs/E8OtNbU

**Comment by Schmazda** [comment]
It actually looks like Velocity does **not** support the Fabric-API properly. Fundamental additions made to the Login Process are not supported by Velocity and are causing these errors. Essentially Velocity has to extend their Login to support said additions made by fabric. Otherwise some crucial packets are either corrupted or not forwarded at all

**Comment by newtykins** [comment]
At least in my case, the issue was that owo-lib (a mod included in Prominence II) has a special handshake protocol which is not supported by Velocity. This was fixed by installing [their Velocity plugin](https://github.com/wisp-forest/owo-velocity-plugin) to add support. I will check whether this is also the case with Dogcraft.

**Comment by isitreallyalive** [comment]
My issue revolved around [a mod](https://github.com/wisp-forest/owo-lib) implementing a custom handshake protocol which Velocity did not understand. After I added the [compat plugin](https://github.com/wisp-forest/owo-velocity-plugin), this error amongst others disappeared. I do not know if Dogcraft has any mods which also implement custom protocols, but it might be worth investigating.

**Comment by Schmazda** [comment]
Yeah, it could be possible to generalize this approach so that every channel gets forwarded like that, but I'm not sure. Either way this would be a temporary workaround and by no means a fix that should be used in production.

**Comment by Schmazda** [comment]
LandonDeam's Problem is probably that (Login-)PluginMessages with the Channel-ID `fabric-networking-api-v1:early_registration` are not forwarded properly which causes kind of a chain reaction that results in non-joinable Fabric servers when the Fabric Registry's size reaches a certain threshold. This should at least apply to the Packet-Size problem.

**Comment by electronicboy** [comment]
owo handshaking occured during the LOGIN phase, a connection phase in which we cannot force the client back into; their compat plugin basically just makes the proxy cache the handshake that occurs during that phase in order to allow it to "support" the thing. blindly resending stuff during that phase is not a solution in which is anything but a hack likely to induce more problems than it solves, thankfully, 1.20.2 has a better solution for that issue.

The entire login phase is generally borked, because we assume that we don't need to have a client and a server talking to one another, given that it's literally impossible to even allow that to happen past the first server, it's never been a big consideration until attempting to design a means to reset modded clients into reconnecting came into play

**Comment by HeberonYT** [comment]
Hi, this topic is old, but it is still a problem for mod communities that use Minecraft 1.20.1, it is the version with the largest variety of mods available.
I have tried Velocity, Waterfall, Lightfall and none of them are able to support a connection with a large forge modpack.
The “Packet Fixer” mod solves many packet problems and works perfectly if I connect to the backend without a proxy.
When using a proxy, I cannot enter my server.
Waterfall will throw the error “io.netty.handler.codec.CorruptedFrameException: length wider than 21-bit”
Velocity does not say the error, it only says that I was disconnected.
It is evident that the problem is the packets that go through the Proxy, is there something to increase the allowed size in the proxy?
Or some proxy that supports modern forge and does not have this communication problem?
My main need is to make a network of 3 servers with the same mods, but we need to move between them.

**Comment by virgil698** [comment]
Is this issue still being discussed by anyone?
I recently ran into a similar situation: when a 1.20.1 Fabric client connects directly to a Fabric server everything works fine, but accessing the server through Velocity throws the following error: https://mclo.gs/IcCLtf4

**Comment by electronicboy** [comment]
That's a separate issue, that looks like you have a mod that broke the vanilla protocol in a manner wr can't support

**Comment by virgil698** [comment]
> That's a separate issue, that looks like you have a mod that broke the vanilla protocol in a manner wr can't support

Does Velocity have any corresponding configuration that supports mods modifying changes to the vanilla protocol?

**Comment by electronicboy** [comment]
No, we can't talk a protocol we don't understand

**Comment by donnie58744** [comment]
Just encountered this today with the mod **Roughly Enough Items** `Packet sent for class com.velocitypowered.proxy.protocol.packet.PluginMessagePacket was too big (expected 229378 bytes, got 635406 bytes)`

**Comment by electronicboy** [comment]
You can use `-Dvelocity.max-plugin-message-payload-size=` to set the maximum allowed payload size, this is now constrained to the vanilla protocols maximum size by default  (and is entirely unrelated to this issue)

**Comment by donnie58744** [comment]
> You can use `-Dvelocity.max-plugin-message-payload-size=` to set the maximum allowed payload size, this is now constrained to the vanilla protocols maximum size by default (and is entirely unrelated to this issue)

You're the man thank you so much fixed it immediately I dont know how I missed that flag. I appericiate the help and fast response 😄
